File: 3443-maximum-manhattan-distance-after-k-changes/3443-maximum-manhattan-distance-after-k-changes.py
class Solution:
    def maxDistance(self, s: str, k: int) -> int:
        curr_n_count = curr_s_count = curr_w_count = curr_e_count = 0
        answer = 0
        for char in s:
            if char == "N":
                curr_n_count += 1
            
            elif char == "E":
                curr_e_count += 1
            
            elif char == "S":
                curr_s_count += 1
            
            else:
                curr_w_count += 1

            max_change = min(curr_e_count, curr_w_count) + min(curr_n_count, curr_s_count)
            unchanged = abs(curr_e_count - curr_w_count) + abs(curr_n_count - curr_s_count)
            answer = max(answer, unchanged + 2 * min(k, max_change))
        
        return answer

File: 3443-maximum-manhattan-distance-after-k-changes/3443-maximum-manhattan-distance-after-k-changes.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var maxDistance = function(s, k) {
    let e_count = w_count = n_count = s_count = 0;
    let answer = 0;
    for (char of s){
        if (char == "S")
            s_count += 1;
        else if (char == "W")
            w_count += 1;
        else if(char == "N")
            n_count += 1;
        else
            e_count += 1;

        let unchanged = Math.abs(s_count - n_count) + Math.abs(e_count - w_count);
        let max_change = Math.min(s_count, n_count) + Math.min(e_count, w_count);
        answer = Math.max(answer, unchanged + 2 * Math.min(k, max_change));
    }

    return answer;
};